cabal.project

```diff
@@ -81,30 +81,61 @@ constraints:
     -- that dependency
   , network >= 3.1.1.0
   , HSOpenSSL >= 0.11.7.2
-
-  -- CONSTRAINTS FOR UPSTREAM CARDANO PACKAGES
-  -- These should all be constraints in upstream packages, but are currently missing.
-  -- We should a) fix these upstream and b) revise the constraints  on the released
-  -- version in CHaP. Unfortunately revisions in CHaP don't work yet, revisit this
-  -- when we fix it.
-
-  -- cardano-prelude: newer protolude changes some signatures in a breaking way, fixed
-  -- in newer cardano-prelude which we can't use yet
-  , protolude < 0.3.1
-  -- cardano-crypto-wrapper: broken by cardano-prelude-0.1.0.1 which moves HeapWords out
-  , cardano-prelude < 0.1.0.1
-  -- cardano-ledger-byron: broken by cardano-binary-1.5.0.1 which includes some functions 
-  -- that used to be in cardano-prelude
-  , cardano-binary < 1.5.0.1
-  -- ledger: cardano-crypto-class-2.0.0.1 is missing some instances needed by the ledger
-  , cardano-crypto-class < 2.0.0.1
-  -- cardano-crypto-tests: we required cardano-crypto-class-2.0.0.0 above, which is broken with
-  -- newer cardano-crypto-tests
-  , cardano-crypto-tests < 2.0.0.1
-  -- plutus-core 1.0.0.1 has the fixed SECP primitives which relies on the newer cardano-crypto-class
-  , plutus-core < 1.0.0.1
-  -- plutus-core: needs a constraint here, fixed on plutus master but not in the released version
   , algebraic-graphs < 0.7
+  , protolude >= 0.3.2
+
+  -- TODO: these should be set in cabal files, but avoiding setting them in lower dependencies for initial CHaP release
+  , cardano-prelude == 0.1.0.1
+  , base-deriving-via == 0.1.0.0
+  , cardano-binary  == 1.5.0
+  , cardano-binary-test == 1.3.0
+  , cardano-crypto-class  == 2.0.0.1
+  , cardano-crypto-praos  == 2.0.0.0.1
+  , cardano-crypto-tests  == 2.0.0.0.1
+  , cardano-slotting  == 0.1.0.0
+  , measures == 0.1.0.0
+  , orphans-deriving-via == 0.1.0.0
+  , cardano-strict-containers == 0.1.0.1
+  , plutus-core == 1.0.0.1
+  , plutus-ledger-api == 1.0.0.1
+  , plutus-tx == 1.0.0.0
+  , plutus-tx-plugin == 1.0.0.0
+  , prettyprinter-configurable == 0.1.0.0
+  , plutus-ghc-stub == 8.6.5
+  , word-array == 0.1.0.0
+  , io-classes >= 0.2.0.0
+  , typed-protocols == 0.1.0.1
+  , typed-protocols-examples == 0.1.0.1
+  , ekg-forward >= 0.2
+
+  , cardano-ledger-alonzo == 0.1.1.1
```

Comment:
Please can we start putting as many of these in the `.cabal` files as we can? I think we should only need these for transitive dependencies where our direct dependencies don't constrain them properly. But I think a lot of these are direct deps and we could just have proper bounds!

Reply:
I can try to push the constraints to package level

Reply:
Unless @angerman wants to keep that for later once we've bumped everything to GHC 9.2?

Reply:
I think it would be good to establish a norm of "any time you go to put a constraint in the cabal.project file or edit one, check to see if it can go in a cabal file", and I don't see why we shouldn't start now :)

Reply:
I got rid of most of the constraints in `cabal.project` and everything seems to build just fine.

Reply:
Still better to put them in the cabal files, at least the major version constraint. That will help us a) avoid picking the next major version when it's released and b) giving people a proper solver error if they don't have a CHaP with the version we need yet, rather than just blindly picking an old version. Bounds are great!